pkg/autoupdate/tuf/library_manager.go

```diff
@@ -0,0 +1,389 @@
+			return nil, errors.New("osquery version query did not return version")
+		}
+
+		currentVersion, err := semver.NewVersion(rawVersion)
+		if err != nil {
+			return nil, fmt.Errorf("could not parse current running version %s of osquery as semver: %w", rawVersion, err)
+		}
+
+		return currentVersion, nil
+	default:
+		return nil, fmt.Errorf("cannot determine current running version for unexpected binary %s", binary)
+	}
+}
+
+// TidyLibrary removes unneeded files from the staged updates and updates directories.
+func (ulm *updateLibraryManager) TidyLibrary() {
+	for _, binary := range binaries {
+		// First, remove old staged archives
+		ulm.tidyStagedUpdates(binary)
+
+		// Get the current running version to preserve it when tidying the available updates
+		var currentVersion *semver.Version
+		var err error
+		switch binary {
+		case binaryOsqueryd:
+			// The osqueryd client may not have initialized yet, so retry the version
+			// check a couple times before giving up
+			osquerydVersionCheckRetries := 5
+			for i := 0; i < osquerydVersionCheckRetries; i += 1 {
+				currentVersion, err = ulm.currentRunningVersion(binary)
+				if err == nil {
+					break
+				}
+				time.Sleep(1 * time.Minute)
+			}
+		default:
+			currentVersion, err = ulm.currentRunningVersion(binary)
+		}
+
+		if err != nil {
+			level.Debug(ulm.logger).Log("msg", "could not get current running version", "binary", binary, "err", err)
+			continue
+		}
+
+		ulm.tidyUpdateLibrary(binary, currentVersion)
+	}
+}
+
+// tidyStagedUpdates removes all old archives from the staged updates directory.
+func (ulm *updateLibraryManager) tidyStagedUpdates(binary autoupdatableBinary) {
+	matches, err := filepath.Glob(filepath.Join(ulm.stagedUpdatesDirectory(binary), "*"))
+	if err != nil {
+		level.Debug(ulm.logger).Log("msg", "could not glob for staged updates to tidy updates library", "err", err)
+		return
+	}
+
+	for _, match := range matches {
+		if err := os.Remove(match); err != nil {
+			level.Debug(ulm.logger).Log("msg", "could not remove staged update when tidying update library", "file", match, "err", err)
+		}
+	}
+}
+
+// tidyUpdateLibrary reviews all updates in the library for the binary and removes any old versions
+// that are no longer needed. It will always preserve the current running binary, and then the
+// two most recent valid versions. It will remove versions it cannot validate.
+func (ulm *updateLibraryManager) tidyUpdateLibrary(binary autoupdatableBinary, currentRunningVersion *semver.Version) {
+	if currentRunningVersion == nil {
+		level.Debug(ulm.logger).Log("msg", "cannot tidy update library without knowing current running version")
+		return
+	}
+
+	const numberOfVersionsToKeep = 3
+
+	rawVersionsInLibrary, err := filepath.Glob(filepath.Join(ulm.updatesDirectory(binary), "*"))
+	if err != nil {
+		level.Debug(ulm.logger).Log("msg", "could not glob for updates to tidy updates library", "err", err)
+		return
+	}
+
+	removeUpdate := func(v string) {
```

Comment:
I don't think it's increasing readability -- I think I could also combine it with the inline `removeBrokenUpdateDir` above in `moveVerifiedUpdate`, since they're basically doing the same thing. I'll do that 💪